### ISSUE TITLE
Set route aliases in the route definition instead of using Yii::setAlias

### ISSUE THREAD

**Issue by yurikosan** [yiisoft/yii2#10414]
I think would be better to set alias in the route structure in order to keep everything together.
http://www.yiiframework.com/doc-2.0/guide-runtime-routing.html#url-suffixes (at the end of paragraph).

```
'rules' => [
    [
        'pattern' => 'posts',
        'route' => 'post/index',
        'alias' => 'posts',
    ],
], 
```

Otherwise, I have to write `\Yii::setAlias('')` for every route.

**Comment by samdark** [comment]
Yep. Could be implemented.

**Comment by klimov-paul** [comment]
Why do you need an alias for URL route?
For me, this sounds like abusing the alias feature.
URL rules serve for URL creation. Any URL for particular controller/action should be composed using `createUrl()`, not by `Yii::getAlias()`.

**Comment by SilverFire** [comment]
Agree with @klimov-paul 
Proposed change increases coupling and might be implemented as standalone extension

**Comment by samdark** [comment]
There's one use case where using route aliases (also known as named routes) are helpful. It's chaning controller names or action names or moving action into another controller. Of course, you can find-replace `['oldcontroller/oldaction'` everywhere but I was asked multiple times about aliases because people want to refactor controllers and actions w/o changing `createUrl` params.

Personally I won't use it.

**Comment by yurikosan** [comment]
Agree with @samdark , exactly for this reason.

**Comment by klimov-paul** [comment]
>There's one use case where using route aliases (also known as named routes) are helpful. It's chaning controller names or action names or moving action into another controller.

Sounds twisted. The result program code will be unreadable: there will be no way to determine which controller particular `createUrl()` actually refers to.
I am against of adding such feature to the core.

**Comment by samdark** [comment]
Yes, of course, it adds extra step when debugging: going to rules list and checking which rule has an alias. That's why I won't use it myself but it seems people want it and are OK with this extra lookup. @yurikosan is that the case?

**Comment by samdark** [comment]
@klimov-paul the thing is that it's already possible and the change proposed makes it easier for people who already use it.

**Comment by yurikosan** [comment]
@klimov-paul you are right only if use of aliases would by mandatory. And why you are not against of using aliases declared via ``` \Yii::setAlias('') ``` ?
@samdark, @klimov-paul OK, It is possible to live without it, it's just convenient, "dirty" habit from Symfony.

**Comment by samdark** [comment]
Let's wait for more people to ask for it.

**Comment by githubjeka** [comment]
>Any URL for particular controller/action should be composed using createUrl(), not by Yii::getAlias()
+1

P.S. PHPStorm can help you to autocomplete for alias name in arrays:
`Yii::$app->params['your-alia(CTRL+SPACE)'];` . For this reason, I prefer to use `params` instead `getAlias`

**Comment by yurikosan** [comment]
@githubjeka I meant this use case from http://www.yiiframework.com/doc-2.0/guide-runtime-routing.html#creating-urls :
```
echo Url::to(['@posts']);
```
It is the only way to aliasing routes and it's from an official manual. 
Therefore route named as '@posts' can be used only after setting Yii::setAlias('@posts', 'posts/index') (e.g), isn't it?

So what's the difference between adding ``` Yii::setAlias('@posts', 'posts/index') ``` and additional element of route's array?

**Comment by githubjeka** [comment]
For me it is strange behavior at least.

**Comment by SilverFire** [comment]
@githubjeka using params for this case seems to me a great idea, as for me.

**Comment by quantum13** [comment]
I agree, that if framework has oportunity of using aliases in Url::to it will be more convenienty to have possibility to set alias in rule.

ps for example in django exist such possibility: https://docs.djangoproject.com/en/1.9/topics/http/urls/#examples

**Comment by samdark** [comment]
@quantum13 it's already supported.

**Comment by quantum13** [comment]
@samdark did you mean, that key 'alias' already supported in rule array and it didn't noted in this topic?

**Comment by samdark** [comment]
No. Using aliases in `Url::to` is supported.

**Comment by quantum13** [comment]
@samdark yes, I know it. I talked about key 'alias' in rule array. Maybe I didn't say clearly anough..

**Comment by samdark** [comment]
Won't be implemented in 2.0.
